doc/developer/chips/18.rst

```diff
@@ -0,0 +1,825 @@
+
+Local access with this technique should be as simple and thus as fast as
+with single-ddata.  Remote access, or actually any access by the NIC
+itself such as the local store of the result of a remote GET, would
+require the additional page translation operation in the comm layer and
+thus be slightly slower.
+
+It would appear that this design has at least some of the same issues
+with mismatches between DSIUtil subdomain partitioning and page-based
+array storage chunking as is the case for regular single-ddata.
+
+For now this awaits more design work.
+
+
+Multi-ddata
+'''''''''''
+
+**Synopsis**: multiple ddatas, individually localized
+
+**Performance snapshot**
+
+    ==========================  ========
+    execution/storage affinity  best
+    access speed                bad/fair
+    balance                     best
+    leverage locality           good
+    network cooperation         good
+    ==========================  ========
+
+
+Description
+...........
+
+Multi-ddata differs from the other techniques in that it allocates
+multiple ddatas, one localized to each NUMA domain, and stores array
+elements on those ddata chunks in the same DSIUtil-based way that the
+``DefaultRectangular`` domain's iterators create partitioning
+subdomains.
+
+
+Analysis
+........
+
+On the positive side, multi-ddata can always get data localization and
+execution/storage affinity right at a reasonable cost, and load-balance
+well.  Assuming we can get NUMA awareness in NIC-registered heaps and
+NUMA-aware allocation, both of which build in a straightforward way on
+NUMA page-placement syscalls which are also needed for localizing
+single-ddata arrays, localization and affinity follow quite simply.
+
+If it needs to get more than one ddata chunk, multi-ddata takes longer
+to allocate space for an array than single-ddata does.  But on the other
+hand it can take advantage of memory that is already localized and also
+memory that is already NIC-registered, when those are available.
+
+What multi-ddata lacks most is access performance.  Array element
+accesses are slow due to extra computations and extra loads of the
+metadata fields used in those computations.  For one thing, computing a
+chunk index from an array index requires doing a divide, effectively.
+This could be sped up significantly by doing an integer multiply by the
+reciprocal instead, or even a right-shift for power-of-two divisors.
+The ddatas and other per-chunk information are themselves currently
+stored as a ddata of records, which could be a tuple of records at a
+fair saving at array creation time.  And the are some other small
+optimizations that could be made in the access code.  Nevertheless while
+the multi-ddata access code could be faster than it is now, it will not
+ever be as simple or fast as plain base-plus-offset addressing.  AT
+least with all the addressing exposed, getting within something like 3x
+of single-ddata is probably the best we can hope for.
```

Comment:
Oh!  Sorry, misunderstood.  Yes, as the number of dimensions rises the linearization down to an offset in the data storage gets more expensive, and the cost of the chunk computation for multi-ddata is a smaller proportion of that because the chunking is on a specific single dimension and thus is a fixed cost.  That not actually an area we've investigated deeply at all, performance-wise.  The current divide is quite expensive compared even to multiplies.  But if we can implement the divide as a multiply-by-reciprocal then indeed multi-ddata should hurt less the more dimensions there are.  I think in the grand scheme of things this is kind of secondary (one more instance of "but in this somewhat-special case multi-ddata doesn't hurt as much"), but I'll add a sentence about it.